File: src/main/java/com/pariveda/kafka/DataEventConsumer.java
package com.pariveda.kafka;

import avro.models.DataEvent;
import com.pariveda.kafka.common.ConfigurationWrapper;
import com.pariveda.kafka.common.StatsDClientFactory;
import com.pariveda.kafka.metrics.Metrics;
import com.timgroup.statsd.StatsDClient;
import org.apache.commons.configuration2.ex.ConfigurationException;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.TopicPartition;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.management.MBeanServer;
import javax.management.ObjectName;
import java.lang.management.ManagementFactory;
import java.util.ArrayList;
import java.util.Properties;

public class DataEventConsumer {
	private static final String CONSUMER_PROPERTIES = "consumer.properties";
	private static final String JMX_METRICS_OBJECT_NAME = "com.pariveda.kafka.metrics:type=ConsumerMetrics";

	private static final Logger log = LoggerFactory.getLogger(DataEventProducer.class);
	private static final MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
	private static final Metrics metrics = new Metrics();

	private static ConfigurationWrapper config;
	private static StatsDClient statsd;

	public static void main(String[] args) {
		if (args.length < 2) {
			System.out.println("Usage: java DataEventConsumer <topic> <partition #>");
			System.exit(1);
		}

		try {
			config = new ConfigurationWrapper(CONSUMER_PROPERTIES);
			statsd = (new StatsDClientFactory(config)).getStatsDClient();
		} catch (ConfigurationException ex) {
			log.error("Configuration exception occurred: {}", ex.getMessage());
			System.exit(1);
		}

		try {
			ObjectName metricsName = new ObjectName(JMX_METRICS_OBJECT_NAME);
			mbs.registerMBean(metrics, metricsName);
		} catch (Exception ex) {
			log.error("JMX exception occurred: {} ", ex.getMessage());
			System.exit(1);
		}

		Properties props = config.getPropertiesFromNamespace("kafka");

		String topic = args[0];
		int partition = Integer.parseInt(args[1], 10);

		ArrayList<TopicPartition> topicPartitions = new ArrayList<>();
		topicPartitions.add(new TopicPartition(topic, partition));

		try (KafkaConsumer<String, DataEvent> consumer = new KafkaConsumer<>(props)) {
			consumer.assign(topicPartitions);
			consumer.seekToBeginning(topicPartitions);

			while (true) {
				ConsumerRecords<String, DataEvent> messages = consumer.poll(10);
				processMessages(messages);
			}
		} catch (Exception e) {
			e.printStackTrace();
			throw e;
		}
	}

	private static void processMessages(ConsumerRecords<String, DataEvent> messages) {
		for (ConsumerRecord<String, DataEvent> message : messages) {
			System.out.println(
					String.format(
							"Topic: %s, Partition: %s, Offset: %s, Key: %s, Value: %s",
							message.topic(),
							message.partition(),
							message.offset(),
							message.key(),
							message.value()));
		}
	}
}

File: src/main/java/com/pariveda/kafka/DataEventConsumer.kt
package com.pariveda.kafka

import avro.models.DataEvent
import org.apache.kafka.clients.consumer.ConsumerRecords
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.TopicPartition
import java.util.*

fun main(args: Array<String>) {
    val props = Properties().apply {
        put("bootstrap.servers", "35.162.160.212:9092,35.162.87.170:9092")
        put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer")
        put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer")
        put("value.deserializer", "com.pariveda.kafka.serialization.DataEventAvroDeserializer")
    }

    if(args.size < 2) {
        System.out.println("Usage: java AvroConsumer <topic> <partition #>")
        System.exit(1)
    }

    val topic = args[0]
    val partition = args[1].toInt()

    val topicPartitions = listOf(TopicPartition(topic, partition))

    try {
        KafkaConsumer<String, DataEvent>(props).use { consumer ->
            consumer.assign(topicPartitions)
            consumer.seekToBeginning(topicPartitions)

            while (true) {
                val messages = consumer.poll(10)
                processMessages(messages)
            }
        }
    } catch(e: Exception) {
        e.printStackTrace()
        throw e
    }
}

private fun processMessages(messages: ConsumerRecords<String, DataEvent>) =
        messages.forEach {
            System.out.println(
                    "Topic: ${it.topic()}, " +
                    "Partition: ${it.partition()}, " +
                    "Offset: ${it.offset()}, " +
                    "Key: ${it.key()}, " +
                    "Value: ${it.value()}")
        }